A phased, diploid assembly of the Cascade hop (Humulus lupulus) genome reveals patterns of selection and haplotype variation

ABSTRACT

Hop (Humulus lupulus L. var Lupulus) is a diploid, dioecious plant with a history of cultivation spanning thousands of years. Hop cones are valued for their use in brewing, and around the world, hop has been used in traditional medicine to treat a variety of ailments, including anxiety, insomnia, and pain. Efforts to uncover the regulation of the biochemical pathways responsible for these desirable traits have been challenged by the large, repetitive, and heterozygous genome of hop. Haploid, short-read assemblies are fragmented and incomplete, containing collapsed regions that thwart the investigation of repeats, genes embedded within repeats, gene copy number, and heterozygosity. We present the first report of a haplotype-phased assembly of the hop cultivar Cascade using PacBio long-read sequencing. Assembly was performed with FALCON, and phasing with FALCON-Unzip. It is the most extensive assembly and gene annotation of the hop genome to date. 

Sequencing and assembly with long-reads improves the resolution of long terminal retrotransposons (LTRs) and captures gene space more comprehensively. Prediction and identification of genes involved in disease resistance and biosynthesis, as well as detection of gene copy number, is vastly improved. Comparative analysis of haplotypes provides insight into large-scale structural variation and lends insight into the function and selective pressures that have driven hop evolution.  We have developed approaches to analyze a diploid genome, including visualization of both haplotypes simultaneously on a synteny browser using GBrowse_syn. These approaches serve to deepen our understanding of the genomic landscape of hop and could have broader applicability to the study of other large, complex genomes. 
